pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/GrowablePriorityLongPairQueue.java

```diff
@@ -0,0 +1,360 @@
+
+/**
+ * An unbounded priority queue based on a min heap where values are composed of pairs of longs.
+ *
+ * When the capacity is reached, data will be moved to a bigger array.
+ * 
+ * <b>It also act as a set and doesn't store duplicate values if {@link #allowedDuplicate} flag is passed false</b>
+ * 
+ ** <p>
+ * (long,long)
+ * <p>
+ *
+ */
+public class GrowablePriorityLongPairQueue {
+
+    private long[] data;
+    private int capacity;
+    private volatile int size = 0;
+    private static final long EmptyItem = -1L;
+
+    public GrowablePriorityLongPairQueue() {
+        this(64);
+    }
+
+    public GrowablePriorityLongPairQueue(int initialCapacity) {
+        checkArgument(initialCapacity > 0);
+        this.capacity = MathUtil.findNextPositivePowerOfTwo(initialCapacity);
+        data = new long[2 * capacity];
+        fillEmptyValue(data, 0, data.length);
+    }
+
+    public interface LongPairPredicate {
+        boolean test(long v1, long v2);
+    }
+
+    public static interface LongPairConsumer {
+        void accept(long v1, long v2);
+    }
+
+    public synchronized void add(long item1, long item2) {
+
+        if (this.size >= this.capacity) {
+            expandArray();
+        }
+
+        int lastIndex = this.size << 1;
+        data[lastIndex] = item1;
+        data[lastIndex + 1] = item2;
+
+        int loc = lastIndex;
+
+        // Swap with parent until parent not larger
+        while (loc > 0 && compare(loc, parent(loc)) < 0) {
+            swap(loc, parent(loc));
+            loc = parent(loc);
+        }
+
+        this.size++;
+
+    }
+
+    public synchronized void forEach(LongPairConsumer processor) {
+        int index = 0;
+        for (int i = 0; i < this.size; i++) {
+            processor.accept(data[index], data[index + 1]);
+            index = index + 2;
+        }
+    }
+
+    /**
+     * @return a new list of all keys (makes a copy)
+     */
+    public Set<LongPair> items() {
+        Set<LongPair> items = new HashSet<>();
+        forEach((item1, item2) -> items.add(new LongPair(item1, item2)));
+        return items;
+    }
+
+    /**
+     * @return a new list of keys with max provided numberOfItems (makes a copy)
+     */
+    public Set<LongPair> items(int numberOfItems) {
+        Set<LongPair> items = new HashSet<>();
+        forEach((item1, item2) -> {
+            if (items.size() < numberOfItems) {
+                items.add(new LongPair(item1, item2));
+            }
+        });
+
+        return items;
+    }
+
+    /**
+     * Removes all of the elements of this collection that satisfy the given predicate.
+     * 
+     * @param filter
+     *            a predicate which returns {@code true} for elements to be removed
+     * @return {@code true} if any elements were removed
+     * 
+     * @return number of removed values
+     */
+    public int removeIf(LongPairPredicate filter) {
+        int removedValues = 0;
+        for (LongPair item : items()) {
```

Comment:
actually, for every delete : heap gets readjusted. so, first we need to know which items satisfy predicate and then remove them one by one (traversing on underlying array and deleting items same time mess up with traversing indexing as heap gets readjusted with new index). 

however, I agree `items()` copies all the items which can be improved. So, updated with additional array which just keeps list of deleted items and then it deletes from the heap.